up-l2/rpcclient.adoc

```diff
@@ -22,79 +22,199 @@ SPDX-License-Identifier: Apache-2.0
+[source]
+----
+invokeMethod(request: UMessage, handler: ResponseHandler)
+----
 
-* Every link:../upclient.adoc[uPClient library] *MUST* implement the `RpcClient` interface.
+.Parameters
+[#parameters, width="100%",cols="15%,25%,60%"]
+|===
+
+| Parameter | Type | Description
+
+| request
+| link:../basics/umessage.adoc[UMessage]
+| The RPC Request message to send to the service provider.
+
+| handler
+| <<response-handler, ResponseHandler>>
+| The handler to invoke with the outcome of the RPC call.
+|===
 
-* The `UUri` and `CallOptions` passed to `invokeMethod()` *MUST* be cached non-persistent memory per call, this is to be able to complete the `Future` when the response is received from the server.
+RpcClient implementations
 
-* The `UPayload` passed to `invokeMethod()` *MAY* be null or empty.
+* *MUST* fail the given handler with a `UCode.DEADLINE_EXCEEDED` if no corresponding Response message has been received before the Request message expires.
+* *MUST* fail the given handler with a `UCode.INVALID_PARAMETER` if the given message is not a link:../basics/uattributes.adoc[valid RPC Request message].
+* *MUST* fail the given handler with a `UCode.ALREADY_EXISTS` if a service invocation with the same request message is currently pending and not expired.
 
-* Method invocation that fails, *MUST* return exceptionally as defined by the link:../basics/error_model.adoc[uProtocol Error Model].
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+sequenceDiagram
+
+actor C as Client
+
+create participant H as handler:ResponseHandler
+C->>H : new()
+participant R as RpcClient
+C-)R : invoke(request, handler)
+activate R
+alt invalid request message
+R--)H : onError(UStatus(UCode.INVALID_ARGUMENT))
+else duplicate request
+R--)H : onError(UStatus(UCode.ALREADY_EXISTS))
+else
+actor S as service provider
+R-)S : request message
+alt request timed out
+R--)H : onError(UStatus(UCode.DEADLINE_EXCEEDED))
+else comm_status != OK
+S-)R : response message
+R--)H : onError(UStatus(comm_status))
+else comm_status = OK
+S-)R : response message
+R--)H : onResponse(UMessage)
+end
+end
+deactivate R
+----
 
-* Cached requests must be completed exceptionally with  `DEADLINE_EXCEEDED` when `UAttributes.isExpired()` is `true`. This notifies the calling context that the request has expired.
+=== Implementation
 
-* *MUST* return `INVALID_PARAMETER` if `UUri` is null or empty or `CallOptions.priority() < CS4`
+Each link:../languages.adoc[language-specific uProtocol library] *MUST* provide a _default_ implementation of `RpcClient` based on the link:../up-l1/README.adoc[Transport Level API] as follows:
 
-* *MUST* return `ALREADY_EXISTS` if the same request already exists (i.e. same `UUri` and `CallOptions`). This is to prevent duplicate requests.
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+sequenceDiagram
+
+actor C as Client
+
+create participant H as handler<br>:ResponseHandler
+C->>H : new()
+participant R as transport based impl<br>:RpcClient
```

Comment:
Given what you've outlined of having a default implementation within the language specific library, how did you see creating an `RpcClient` looking like?

Would that default implementation take a reference to a `UTransport` implementation?

e.g.

```rust
struct DefaultRpcClient {
  transport: Arc<dyn UTransport>
}

impl DefaultRpcClient {
  pub fn build(transport: Arc<dyn UTransport>) -> Arc<dyn RpcClient> {
    // ...
  }
}
```